Board game setup bd

BlakeDonn opened the pull request:
### PR Template:
#### What's this PR do?  

- Adds board_game schema, seed, and datafile
- Seeds boardgame data through the boardgameatlas API instead of having to hardcode
- Easily dynamic to adjust for more/less or different boardgames( check out the games array, its as simple as just adding names to that)

#### Where should the reviewer start?  

- Start by checking out the logic on the data population to see if it can be cleaned up at all

#### How should this be manually tested?  

- Pull down the local changes
- Run `knex migrate:latest` and `knex seed:run` to populate the databse
- Check out the data that is entered now under the board_games table through Postico or other means

#### Any background context you want to provide?  
- The API doesn't have a crazy amount of information, but if we are hindering on someone already owning a boardgame do we really need to provide much other than the basics?
- Some broken data for a couple of the games, may need to filter those out in favor of ones who have all the data
- Feel free to adjust any of the boardgame names, pretty much just tried to find the most popular ones I could think of
 
#### What are the relevant tickets?  
 closes #10 

#### Screenshots (if appropriate) 

![Screen Shot 2020-10-28 at 11 47 33 PM](https://user-images.githubusercontent.com/65369275/97523183-f5c40900-1977-11eb-8b1e-647d872bfdf1.png)

BlakeDonn committed:
Add movies migration

BlakeDonn committed:
Add movie API calls

BlakeDonn committed:
Add movie dataset generator function

BlakeDonn committed:
Add movieSeed file

BlakeDonn committed:
Adjust categories to be more family friendly

BlakeDonn committed:
Adjust logic to return 25 family friendly movies

BlakeDonn committed:
Adjust reduce to be async

BlakeDonn committed:
Adjust dev server name

BlakeDonn committed:
Adjust export syntax

BlakeDonn committed:
Remove fetch and move to data file

BlakeDonn committed:
Add promise handler for individual movies

BlakeDonn committed:
Add new properties to seed for movie

BlakeDonn committed:
Finalize safetynets for bad data, remove logs

BlakeDonn committed:
Remove reduce in favor of while loop

BlakeDonn committed:
Adjust rating from integer to string

BlakeDonn committed:
Adjust api key

BlakeDonn committed:
Adjust message while seeding

BlakeDonn committed:
Adjust seeding message to log

BlakeDonn committed:
Remove console log, add return for catches

BlakeDonn committed:
Remove unneeded genres

BlakeDonn committed:
Add movie_plot column

BlakeDonn committed:
Adjust seed to poulate movie_plot

BlakeDonn committed:
Add board_game migration file

BlakeDonn committed:
Add initial structure

BlakeDonn committed:
Reformat to match card_game schema

BlakeDonn committed:
Refactor return on moviecalls, add boardGame fetch

BlakeDonn committed:
Refactor data to more accurately populate data

BlakeDonn committed:
Rename file

BlakeDonn committed:
Refine data to store based on API

BlakeDonn committed:
Refine schema to reflect the data coming in

BlakeDonn committed:
Adjust insert syntax

BlakeDonn committed:
Adjust naming

BlakeDonn committed:
Adjust seed to accurately represent data

BlakeDonn committed:
Remove lingering movie syntax

BlakeDonn commented on db/data/board_games.js:
```diff
@@ -0,0 +1,18 @@
+const apiCalls = require("../../apiCalls.js")
+
+const board_gameData = async () => {
+  const games = ['chess', 'checkers', 'jenga', 'clue', 'scrabble', 'connect four', 'candy Land', 'exploding Kittens', 'scrabble', 'life', 'guess who', 'operation', 'Scattergories', 'apples to Apples', 'monopoly', 'battleship', 'mouse trap', 'trivial Pursuit', 'charades', 'pictionary']
```
Here is where you can feel free to add or remove any names, should work pretty straight forward if you just follow the naming convention.

Atos20 replied:
sorry but what is the actual naming convention here, I noticed `mouse trap` doesn't capitalize the second word but `trivial Pursuit` does.

BlakeDonn replied:
woops! I'll fix that and push it up, good call. I guess overall naming convention is really just a space between anything that has two words. upper/lowercase doesn't seem to make much difference, lets just stick to all lowercase to keep it consistent.

nicolegooden replied:
Good catch, Orlando! I think consistency is key here, so whichever seems to appear most often in the database as of now, should be what we go with.

BlakeDonn commented on db/migrations/20201028220558_board_game.js:
```diff
@@ -0,0 +1,16 @@
+exports.up = knex => {
+  return knex.schema.createTable('board_games', function (table) {
+    table.increments('id').primary()
+    table.string('name')
+    table.string('description', 10000)
+    table.string('min_players')
+    table.string('max_players')
+    table.string('image')
+    table.string('average_time').defaultTo(1)
+    table.timestamps(true, true)
```
Let me know if you have any ideas for any other relevant information we could add

Atos20 replied:
more than an idea for another column, I am still a little bit unsure when to use `table.string` and `table.text`. I think it would be more semantic to use `table.text` when we use content to describe or comment. Let me know what you all think.

What I have found is that we can use `table.text` in this case for line `5`

BlakeDonn replied:
I will try this out and see if the data seeds properly; If it does, i'll push a fix with changing string to text.

nicolegooden replied:
Yeah, here is what I found in my own research:

`The difference between string and text is the intended size: string is typically of shorter, fixed, length, while text implies a longer, variable-length, string.`

So, I'd say anything that is longer than a sentence  or two should probably be `table.text('column_name')`

Atos20 commented on db/seeds/dev/board_gameSeed.js:
```diff
@@ -0,0 +1,24 @@
+  const boardId = await knex('board_games').insert({
+    name: board.name,
+    description: board.description || board.description_preview,
+    min_players: board.min_players,
```
this is such good use of the `or`, I should implement it as well for the music and podcast table, ssuming that `description` and `previews` are values returned by your promise. 💯

BlakeDonn replied:
Yeah the data on some of these APIs seems a little inconsistent as far as I have seen. I used the && (logical and) operator  on the movies api to fix some breaking bugs (checking for nested data where the parent doesn't always exist), so that also seems super helpful!

Atos20 commented on db/seeds/dev/moviesSeed.js:
```diff
@@ -0,0 +1,28 @@
+    release_date: movie.releaseDate,
+    runtime: movie.title.runningTimeInMinutes,
+    where_to_watch: "not sure",
+  }, 'id');
```
could you expand on what line 15 is doing?

BlakeDonn replied:
I was wondering about this too, I took it from the knex express lesson boilerplate. I just tested removing it and it seems like it changed nothing. Going to remove it from both seed files I have used so far and push back. No sense in having code that doesn't do anything.

nicolegooden replied:
@BlakeDonn I wondered the same thing when I saw it in the lesson.  I figured having the `id` in there isn't necessary for the scope of our project.

Atos20 commented on knexfile.js:
```diff
@@ -16,7 +16,7 @@ module.exports = {
     client: 'pg',
-    connection: 'postgres://localhost/school_test',
+    connection: 'postgres://localhost/',
     migrations: {
```
I wonder if we need to incorporate a `testing environment` when testing the Express calls but I am unsure how will this look like. Now that I think of it, this might be out of the scope for this project tho.

BlakeDonn replied:
From the testing lesson it didn't seem like it was much different than just making another database and running commands a little differently.

Atos20 commented on db/data/board_games.js:
```diff
@@ -0,0 +1,18 @@
+const apiCalls = require("../../apiCalls.js")
+
+const board_gameData = async () => {
```
I have been using camelCase to name my function, should we write all our `server-side` code in sanake_case and `front-end` camelCase?

BlakeDonn replied:
I am not sure honestly, this looks a bit contrived to me, and seems like camelCase would be just fine.

nicolegooden replied:
I have been using camelCase for everything except the names of tables and columns for the database, just FYI!

Atos20 reviewed (approved):
this looks really good thanks

BlakeDonn committed:
Standardize naming convention

BlakeDonn committed:
Adjust description chartype to text from string to remove charlimit

BlakeDonn committed:
Remove unneccessary ID from insert function

BlakeDonn committed:
Adjust to have api keys hidden in .env

BlakeDonn committed:
Update dependancies to have dotenv package

BlakeDonn committed:
Remove log

BlakeDonn committed:
Reduce number of movies being seeded to 3 to avoid burning through api key limit

BlakeDonn committed:
Adjust chartype to text to avoid limits on description and plot

BlakeDonn commented:
Made some slight adjustments based on feedback:
1. Naming convention for board games is now consistent "candy land", "battleship", "trivial pursuit" is the convention to follow
2. Changed charType to text for `movie_plot` and `brief_description` to remove character limits
3. Removed API keys from apiCalls for good practice (api key will now live in .env file locally and be accessed through `dotenv` package) <- I'll share the key and relevant setup instructions directly
4. Removed redundant or unused information on inserts

nicolegooden commented:
Nice work! I'll make sure to change the names of card games in the `card_games` table so that everything is lowercase.  I believe right now, every name has the first letter of each word capitalized.

nicolegooden commented on db/migrations/20201028170723_movies.js:
```diff
@@ -0,0 +1,18 @@
+    table.string('release_date')
+    table.string('runtime')
+    table.string('title')
+    table.string('where_to_watch')
```
Should we have `timestamps` on this table like we do for the board games?

nicolegooden commented:
Also, I agree that the board games table shouldn't have too much information.  I think what you have included is perfectly fine.
